### PR TITLE
New rule S6960 for C#: Controllers should not have too many responsibilities

### PR DESCRIPTION
**Pull request by antonioaversa**
First part of #9089

### CODE REVIEW

**Comment by antonioaversa** [comment]
Moved back to draft as the Disjoint Sets primitives have been extracted into a dedicated class and a dedicated PR: 
https://github.com/SonarSource/sonar-dotnet/pull/9145

This PR now has https://github.com/SonarSource/sonar-dotnet/pull/9145 as a dependency.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:144]: `ImplicitlyDeclared` is pulling double-duty here, right?

- Property backers
- Primary constructor parameter backers

I would add a comment to make this a bit more clear:

```csharp
// Exclude backing fields for auto-properties and primary constructor parameters
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:71]: Educational: Since the code doesn't care about order, why not `ConcurrentBag`?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:95]: I think it makes sense for the method to describe what it does (disjoint set algorithm), but the variable should be named according to what it semantically represents. For example, `responsibilityUnits`.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:169]: I think going for an old-fashioned double for-loop is both more readable and more performant here:

```csharp
var methods = controllerSymbol.GetMembers().OfType<IMethodSymbol>().ToList();
for (var setIndex = 0; setIndex < sets.Count; setIndex++)
{
    foreach (var memberLocation in sets[setIndex].SelectMany(MemberLocations))
    {
        yield return new SecondaryLocation(memberLocation, $"Belongs to responsibility #{setIndex}.");
    }
}

IEnumerable<Location> MemberLocations(string memberName) =>
    methods.Where(x => x.Name == memberName).SelectMany(LocationIdentifiers<MethodDeclarationSyntax>);
```

Advantages: 
- You do not need to make an enumerable range just to get the set index indirectly
- You do not need to zip, making more enumerators
- You do not need to use archaic query-syntax incantations (joking, it's a preference and sometimes stronger)
- Semantically, each loop does one of the two things you need:
  - Outer loop gives you setIndex
  - Inner loop gives you location

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:102]: Educational: Why use the recognizer at this level? Why not the simpler `ReportIssue` overload?

**Review by gregory-paidis-sonarsource — CHANGES_REQUESTED** [review]
Amazing work, very clever and clean implementation 💘 
Left some questions and some cleaning-up suggestions.
I did not go over the test cases again, I trust that you copied what was already reviewed on the RSPEC side.
If there are changes, please let me know with a comment where I should look.

File: analyzers/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs:185]: I have added dedicated tests for `IsStaticConstructor`, for both C# and VB.

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs:44]: I went with the first option.

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs:48]: Until the circle is finalized, the outcome of the circle should not be applied to production work.
Let's do such tests in another context, possibly in a branch that doesn't get merged.

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs:47]: See [this comment](https://github.com/SonarSource/sonar-dotnet/pull/9111#discussion_r1572431736).
I wouldn't do cross-experimentations that can possibly affect delivery.
We can do such experiments on a more contained and dedicated context.

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs:33]: When you combine your three suggestions "use a static builder", "initialize the references in the builder" and "use a static field for references" you get a potential initialization issue, since the order of initialization of inline static fields is not guaranteed, and `references` may not be initialized while you are initializing `builder`.

You never fall into such traps with static fields initialization if you use a property. The property is anyway only accessed once, so there is virtually no performance loss once the code is JITed.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:31]: When a collection is used in a multi-threaded environment (as in multiple registrations), you need to have thread safety, so you need to use an immutable/frozen data structure, possibly with a lock-free behavior.
When a collection is used in a single-thread context, such as `RelevantMemberNames`, you don't need such guarantees.
I can add a comment to clarify that.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:71]: Because, unlike `ConcurrentStack`, it's not lock-free ([ref](https://devblogs.microsoft.com/pfxteam/faq-are-all-of-the-new-concurrent-collections-lock-free/)). 
This is where we need to be careful with thread contention. The lambda registering on every symbol identifier may be called by N different threads, and using a lock-free collection ensures no thread contention.
That's a general thing that we have been doing with all rules that use `RegisterSymbolStartAction` and register to syntax nodes in that context, to gather data into a single data structure.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:102]: No reason, copy-paste thing I guess. I have changed that to the single-parameter overload.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:32]: I get the concept in theory, but aren't you using this HS in a multi-threaded context?
It is used in `SymbolStart`, which I assume happens concurrently for all the symbols.

Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:31]: Yes, indeed. There is one difference, though: `ExcludedWellKnownServices` is instantiated statically, and it's never modified afterward. That's an important difference since it pretty much guarantees a memory barrier, which is important for the `Contains` to be de facto thread-safe (in practice, nothing is guaranteed, really, I am afraid—[ref](https://stackoverflow.com/a/28960891)).

Again in practice, if we couldn't assume read-safety of `HashSet<T>` and related, we would need to change a lot of code in our codebase.

On the other hand, `memberNames` is instantiated in `CheckApiController` and used in closures two levels down.
It is true we could use the same argument that `memberNames` is not modified after it's instantiated in `RelevantMemberNames`, so it should be safe as well. However, given the fact that the collection is local, and not static, I am a bit more reluctant to use a thread-unsafe collection here.

If you have arguments in favor, I can change `ImmutableHashSet<T>` to `HashSet<T>`.

**Comment by antonioaversa** [comment]
@gregory-paidis-sonarsource I have undrafted the PR, after changing the base to the PR with the Disjoint Sets data structure. 
You can now go for a second round of review.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:126]: This is a tad bit expensive.
Maybe it's better to extract it to a field and set it once in the initialization?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:157]:
```suggestion
        var methods = controllerSymbol.GetMembers().OfType<IMethodSymbol>();
```
There's no need to evaluate this here - it's going to happen in the loop later anyways/

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:54]:
```suggestion
                        && symbol.GetAttributesWithInherited().Any(x => x.AttributeClass.DerivesFrom(KnownType.Microsoft_AspNetCore_Mvc_ApiControllerAttribute)))
```

Maybe we can extract this into a method `IsApiController` in the helper class?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:126]: What do you mean by this?
`GetMembers()` is called once per `symbol`.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:157]: By adding `ToList` it is evaluated only once, instead of once per `set` inside the `setIndex` outer-loop.

File: analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ControllersHaveTooManyResponsibilitiesTest.cs:47]: I would agree with this in general, but we do not really have a hard-enforced convention about how to structure our UTs.
And I do not see any danger with doing this, at least not until C#13 is released.
I will leave it up to you, since you make a valid argument.

**Review by gregory-paidis-sonarsource — APPROVED** [review]
LGTM, only one open comment remaining from me which is optional

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:157]: Indeed. However, this, from my POV, is confusing the way it's written.
If Anton applies the other comment about extracting the GetMembers in a field:

```
    private static IEnumerable<SecondaryLocation> SecondaryLocations(List<List<string>> sets)
    {
        var membersLocations = MemberLocations(members.OfType<IMethodSymbol>());
        for (var setIndex = 0; setIndex < sets.Count; setIndex++)
        {
            foreach (var memberLocation in sets[setIndex].SelectMany(x => membersLocations[x]))
            {
                yield return new SecondaryLocation(memberLocation, $"Belongs to responsibility #{setIndex + 1}.");
            }
        }

        static IDictionary<string, IEnumerable<Location>> MemberLocations(IEnumerable<IMethodSymbol> members) =>
            members.GroupBy(x => x.Name).ToDictionary(x => x.Key, x => x.SelectMany(LocationIdentifiers<MethodDeclarationSyntax>));
    }
    ```
    
    Also please move the `LocationIdentifiers` method after this since it's a leaf.

**Review by mary-georgiou-sonarsource — CHANGES_REQUESTED** [review]
Polishing.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:126]: Ah that's a fair point, I see what you mean.
I agree with your suggestion, with one caveat:

It cannot be a field, as it should be scoped to the registration callback. 
The analyzer should not have any state.
We could do `var members = symbol.GetMembers();` at the top and pass it to both methods you mentioned.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:126]: Indeed! Thanks @gregory-paidis-sonarsource .

Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs:126]: Roslyn has a [specific API](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.inamespaceortypesymbol.getmembers?view=roslyn-dotnet-4.7.0#microsoft-codeanalysis-inamespaceortypesymbol-getmembers(system-string)) for retrieval by name, that works with overloads.
I have used that one instead.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs]: `ToList` is gone using the name-specific overload of `GetMembers`.

File: analyzers/tests/SonarAnalyzer.Test/Common/DisjointSetsTest.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Common/DisjointSetsTest.cs:1]: Useless using

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/ControllersHaveMixedResponsibilities.Latest.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/ControllersHaveMixedResponsibilities.Latest.cs:868]: I would add a test that looks exactly like this, but derives from `Controller`.
Just to be explicit that we do not take it into account.

**Comment by mary-georgiou-sonarsource** [comment]
For the history of things - I took over the last details of this PR as @antonioaversa had to move to another sprint.

**Review by gregory-paidis-sonarsource — COMMENTED** [review]
Looks good, left only two small suggestions.

I have a couple of questions, as I missed the previous meeting:

1. How come we did not go with the attributes, and instead take into account all classes/interfaces by default?
2. Do you plan to exclude abstract controllers before we merge, or will you do a small update PR just for that?

**Comment by mary-georgiou-sonarsource** [comment]
> Looks good, left only two small suggestions.
> 
> I have a couple of questions, as I missed the previous meeting:
> 
> 1. How come we did not go with the attributes, and instead take into account all classes/interfaces by default?
> 2. Do you plan to exclude abstract controllers before we merge, or will you do a small update PR just for that?

1. We decided to not rush adding the attributes as they cannot be removed once added. It would be problematic for the users.
2. I've added the abstract classes exception.

**Comment by mary-georgiou-sonarsource** [comment]
@gregory-paidis-sonarsource BTW, I'd ignore the code smell as it won't fix for us.

**Pull request merged**
